R/schambach_table.R

```diff
@@ -0,0 +1,157 @@
+  )
+  
+  roster %>% 
+    ensurer::ensure_that(
+      all(subgroup_cols %in% names(roster)) ~ "subgroup_cols must match column names in your mapvizieR roster."
+    )
+  
+  if (complete_obsv == TRUE) {
+    this_growth <- this_growth %>%
+      dplyr::filter(
+        complete_obsv == TRUE  
+      )
+  }
+  
+  group_summary <- function(grouped_df) {
+    
+    df <- grouped_df %>%
+      summarize(
+        end_rit = round(mean(end_testritscore, na.rm=TRUE), digits=1),
+        start_top75 = round(100 * sum(start_testpercentile >= 75) / n(), digits=1),
+        end_top75 = round(100 * sum(end_testpercentile >= 75) / n(), digits=1),
+        avg_pg = round(mean((end_testpercentile - start_testpercentile), na.rm=TRUE), digits=1),
+        p_ku = round(100 * sum(met_typical_growth, na.rm=TRUE) / n(), digits=1),
+        p_rr = round(100 * sum(met_accel_growth, na.rm=TRUE) / n(), digits=1),
+        n = n()
+      ) %>% 
+      as.data.frame
+    
+    df
+  }
+  
+  tables <- list()
+  for (i in 1:length(subgroup_cols)) {
+    subgroup <- subgroup_cols[i]
+    
+    #join roster and data
+    if (subgroup == 'end_schoolname') {
+      minimal_roster <- roster[, c('studentid', 'map_year_academic', 
+                                   'fallwinterspring')]
+    } else {
+      minimal_roster <- roster[, c('studentid','map_year_academic',
+                                   'fallwinterspring','end_schoolname',subgroup)]
+    }
+    
+    combined_df <- dplyr::inner_join(
+      x = this_growth,
+      y = minimal_roster,
+      by = c('studentid' = 'studentid', 
+             'start_map_year_academic' = 'map_year_academic', 
+             'start_fallwinterspring' = 'fallwinterspring')
+    )
+    
+    #first row: summary of entire grade level
+    row1 <- c(
+      paste('Total: Grade',grade),
+      round(mean(combined_df$end_testritscore, na.rm=TRUE), digits=1),
+      round(100 * sum(combined_df$start_testpercentile >= 75) / nrow(combined_df), digits=1),
+      round(100 * sum(combined_df$end_testpercentile >= 75) / nrow(combined_df), digits=1),
+      round(mean((combined_df$end_testpercentile - combined_df$start_testpercentile), na.rm=TRUE), digits=1),
+      round(100 * sum(combined_df$met_typical_growth, na.rm=TRUE) / nrow(combined_df), digits=1),
+      round(100 * sum(combined_df$met_accel_growth, na.rm=TRUE) / nrow(combined_df), digits=1),
+      nrow(combined_df)
+    )
+    
+    #now group by subgroup and summarize
+    grouped_df <- dplyr::group_by_(combined_df, subgroup)
+    this_summary <- group_summary(grouped_df)
+    names(this_summary) <- c(pretty_names[i],'Avg. Ending RIT', 'Percent Started in Top 75%',
+                             'Percent Ended in Top 75%', 'Avg. Percentile Growth',
+                             'Percent Meeting KU', 'Percent Meeting RR','Number of Students')
+    
+    tables[[i]] <- rbind(row1,this_summary)
+    for (c in 2:ncol(tables[[i]])) {
+      tables[[i]][,c] <- as.numeric(tables[[i]][,c])
```

Comment:
what does this line do?  forces numeric columnwise?  why does this happen after it is put on the list?